### PR TITLE
MDEV-28315 Fix ASAN stack-buffer-overflow in String::copy_aligned

### PR DESCRIPTION
**Pull request by rdtr**
Pushed for discussion, I will modify the description later.

- [x] *The Jira issue number for this PR is: MDEV-28315

## Description
Starting since this commit 36cdd5c3cdb06d8538f64c0b312ffe4672a92e75 there is an ASAN stack-buffer-overflow error because we append a NULL terminator beyond the length of memory allocated. According to the comment, after the commit we consolidate the responsibility to append a terminator to `c_ttr()` if necessary therefore appending a NULL terminator here should not be necessary.

## How can this PR be tested?
Added one test case to `main.strings`.

## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [x] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

### CODE REVIEW

File: mysql-test/suite/innodb/t/MDEV-28315.test [hunk 1]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE t1 (a VARBINARY(128)) CHARACTER SET utf32;
+INSERT INTO t1 VALUES ('South Carolina, Vermont, New Jersey, New Mexico, Wisconsin, Missouri, Delaware');
+CREATE TABLE t2 (b SET('South Carolina', 'Vermont', 'Texas', 'New Mexico', 'Wisconsin', 'Missouri', 'Delaware', 'Wyoming', 'New Jersey', 'Maryland', 'Illinois', 'New York')) CHARACTER SET utf32;
+--error WARN_DATA_TRUNCATED
+INSERT INTO t2 SELECT * FROM t1;
```
Comment (nayuta-yanagisawa) [inline mysql-test/suite/innodb/t/MDEV-28315.test:5]: No newline at end of the line

**Review by nayuta-yanagisawa — COMMENTED** [review]
@rdtr Thank you for your pull request. Please add some explanation of the patch to the commit message and the description section of your first comment? From the commit message ...

* One should be able to tell roughly what a commit does without reading the code.
* One should be able to understand the reasoning for a commit from its message.

File: sql/sql_string.h [hunk 2]
```diff
@@ -701,6 +701,9 @@ class Binary_string: public Sql_alloc
   bool realloc_raw(size_t arg_length);
   bool realloc(size_t arg_length)
   {
+    // This increment is redundant because we allocate the size incremented 
+    // inside realloc_raw. Removing this +1, however, introduces a lot of test failures
+    // so we leavet it for now. It does no harm to allocate one extra byte.
```
Comment (dlenski) [inline sql/sql_string.h:706]: *`leave`, not `leavet`

Comment (dlenski) [inline sql/sql_string.h:706]: > Removing this +1, however, introduces a lot of test failures

What kind of test failures?

If this `+1` appears logically unnecessary, but removing it causes failures, that suggests to me that there must be an off-by-`-1` error somewhere in `realloc_raw`, which the `+1` here is mitigating by wrapping it.

**Comment by dlenski** [comment]
Trying to piece together a clear story of the bug and the fix here, from reading [MDEV-28315](https://jira.mariadb.org/browse/MDEV-28315) and this PR.

it appears that:

1. There is a bug in low-level string-handling methods.
2. That bug causes `String::copy_aligned` to sometimes produce a string buffer that does *not* contain expected terminating `NUL` byte.
3. That bug can be triggered by `INSERT`ing an overly long value into a `SET`-type column. (Is that the _only_ way to trigger it?)
4. This bug can be triggered in this way since 36cdd5c3cdb06d8538f64c0b312ffe4672a92e75 (Shown via `git-bisect`? Or what?)
5. This PR addresses that bug by ensuring that `String::copy_aligned` also writes a terminating `NUL` byte.
6. This PR also identifies an apparent, unnecessary `+1` in `Binary_string::realloc`. However, attempting to _remove_ that `+1` leads to "a lot of test failures".
  - This suggests to me that there's an off-by-one error in the _opposite direction_ in the wrapped `Binary_string::realloc_raw` function

Am I understanding all of that right? Is only `Binary_string::realloc` susceptible to this bug? Is `String::realloc` not susceptible?

File: mysql-test/suite/innodb/t/MDEV-28315.test (hunk 1 above)
Comment (dlenski) [reply mysql-test/suite/innodb/t/MDEV-28315.test:5]: There should be a `SELECT * FROM t2` at the end here, to explain and verify the expected contents of `t2` if the memory bug is not present.

I expected the corresponding result would just be:
```
b
South Carolina
```

Comment (rdtr) [reply mysql-test/suite/innodb/t/MDEV-28315.test]: Thanks. I added the statement. The output was just `b` so looks like that `b` is an empty table. I talked with @nayuta-yanagisawa and concluded it is expected as a result of truncation.

File: sql/sql_string.h [hunk 3]
```diff
@@ -690,7 +690,7 @@ class Binary_string: public Sql_alloc
 
       Note that if arg_length == Alloced_length then we don't allocate.
       This ensures we don't do any extra allocations in protocol and String:int,
-      but the string will not be atomically null terminated if c_ptr() is not
+      but the string will not be automically null terminated if c_ptr() is not
```
[the remark this answers is not in the document]
Comment (rdtr) [reply sql/sql_string.h:693]: Fixed. Thanks.

File: sql/sql_string.h (hunk 2 above)
Comment (rdtr) [reply sql/sql_string.h]: Fixed. Thanks.

Comment (rdtr) [reply sql/sql_string.h]:
```
worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 16000..16019
main.pool_of_threads                     [ fail ]
        Test ended at 2022-05-22 14:00:32

CURRENT_TEST: main.pool_of_threads
mysqltest: In included file "./include/common-tests.inc":
included from /home/norakagi/projects/mariadb-dev/server/MDEV-28315/mysql-test/main/pool_of_threads.test at line 12:
At line 13: query 'drop table if exists t1,t2,t3,t4' failed: <Unknown> (5): Out of memory (Needed 24 bytes)

The result from queries just before the failure was:
SET optimizer_switch='outer_join_with_cache=off';
drop table if exists t1,t2,t3,t4;

 - saving '/home/norakagi/projects/mariadb-dev/server/MDEV-28315/bld/mysql-test/var/log/main.pool_of_threads/' to '/home/norakagi/projects/mariadb-dev/server/MDEV-28315/bld/mysql-test/var/log/main.pool_of_threads/'
--------------------------------------------------------------------------
The servers were restarted 0 times
Spent 0.000 of 2 seconds executing testcases

Failure: Failed 1/1 tests, 0.00% were successful.
```

One failure I found is as above. You are correct but that I think should be a separate issue.

Comment (dlenski) [reply sql/sql_string.h]: Agreed. It's a somewhat separate issue.

**Comment by rdtr** [comment]
@[dlenski](https://github.com/dlenski) Replied inline.

> 1. There is a bug in low-level string-handling methods.

It is true.

> 2. That bug causes `String::copy_aligned` to sometimes produce a string buffer that does _not_ contain expected terminating `NUL` byte.
> 3. That bug can be triggered by `INSERT`ing an overly long value into a `SET`-type column. (Is that the _only_ way to trigger it?)
> 4. This bug can be triggered in this way since [36cdd5c](https://github.com/MariaDB/server/commit/36cdd5c3cdb06d8538f64c0b312ffe4672a92e75) (Shown via `git-bisect`? Or what?)

Yes this bug was introduced by that commit according to the ticket. However the issue is different in my understanding. `String::copy_aligned` was trying to always append a NULL terminator but actually the code does not allocate a space for the terminator, therefore it caused stack buffer overflow, which was the issue brought up on the ticket.

As is the comment added in the commit,

      but the string will not be automatically null terminated if c_ptr() is not
      called.

It looks rather the author's intention was not to always add a terminator here and to rely it on `c_ptr()` call. So in my commit, I removed the code to append the terminator. This can potentially introduce another failure if the code has any case that we fail to call `c_ptr()` whenever necessary, but considering that tests all passed I cannot but rely upon this author's comment and existing tests. I expect MariaDB team can take a look to check if all code paths catches up with the optimization introduced by the original commit.

> 5. This PR addresses that bug by ensuring that `String::copy_aligned` also writes a terminating `NUL` byte.

So it is opposite, I made the change so as NOT to append a terminating NULL in the code path.

> 6. This PR also identifies an apparent, unnecessary `+1` in `Binary_string::realloc`. However, attempting to _remove_ that `+1` leads to "a lot of test failures".
> 
> * This suggests to me that there's an off-by-one error in the _opposite direction_ in the wrapped `Binary_string::realloc_raw` function
> 

Yes, I am not sure there are "a lot of" but after spending some time I feel it's more complicated to investigate the new test failure so want to leave the +1 part as it is for now.

**Comment by rdtr** [comment]
Created https://jira.mariadb.org/browse/MDEV-28646 for the redundant +1 allocation.

Also talked with @nayuta-yanagisawa and `SELECT * FROM t2` may not be necessary because this issue is related specifically to String class implementation and as long as we no longer see the assertion failure it should be good (The content of t2 does not matter here).

so I've omitted the part from the test file. Let me know if any comments you have. Thanks !

**Comment by nayuta-yanagisawa** [comment]
@rdtr I had a private discussion with Monty. I now believe that the current code of `realloc()` is correct. In some cases, we may reserve one extra byte that is not needed. However, as he said, it was intentional and he didn't want to add another malloc & realloc code to cover this unlikely case.

He suggested to fixing the problem by allocating additional space in `copy_aligned()` and leave `Ptr[aligned_length]=0` as is:

```
--- a/sql/sql_string.cc
+++ b/sql/sql_string.cc
@@ -398,7 +398,7 @@ bool String::copy_aligned(const char *str, size_t arg_length, size_t offset,
   DBUG_ASSERT(offset && offset != cs->mbminlen);
 
   size_t aligned_length= arg_length + offset;
-  if (alloc(aligned_length))
+  if (alloc(aligned_length+1))
     return TRUE;
   
   /*
```

**Comment by rdtr** [comment]
@nayuta-yanagisawa @montywi The change is aligned to what Monty suggested. Can you check the current revision ?

**Comment by rdtr** [comment]
@nayuta-yanagisawa Thank you for the catch ! It should be good now after verifying the tests passed.

**Review by nayuta-yanagisawa — APPROVED** [review]

**Comment by nayuta-yanagisawa** [comment]
Pushed to bb https://github.com/MariaDB/server/commit/bb3300d60774b10a5b8a84d2c6baf60d10f41e42

**Comment by nayuta-yanagisawa** [comment]
```
MDEV-28315 Fix ASAN stack-buffer-overflow in String::copy_aligned                            

Starting since this commit 36cdd5c there is an ASAN
stack-buffer-overflow error because we append a NULL terminator
beyond the length of memory allocated.

Reviewed by: Monty and Nayuta Yanagisawa
```

**Pull request merged**